kumahq/kuma
# chore(kds): remove kds v1

This PR aims to remove the deprecated KDS v1 source codes.

# Elaboration

## What is KDS?

KDS is Kuma Discovery Service which works for resource synchronization(Policies/Ingress..) between `Global` and `Zone`

### KDS v1 and v2 context

#### v1

- With the [first kds server PR](https://github.com/kumahq/kuma/pull/839), we implemented something like the go-control-plane `SotW` protocol.

- With the [upgrade go-control-plane PR](https://github.com/kumahq/kuma/pull/1290), we use the go-control-plane `SotW` SDK directly.

#### v2

- KDS redesign: https://github.com/kumahq/kuma/issues/4354

- KDS v2 implementation: https://github.com/kumahq/kuma/pull/6278

#### Why did we migrate kds to v2 ?

1. The v1 KDS uses GRPC bidirectional stream which is hard to implement. (We also met a bug with that: https://github.com/kumahq/kuma/pull/5373)
2. The v1 KDS uses go-control-plane `SotW` protocol which can lead to a low efficiency when there are many data that need to be exchanged.

For more details you can read [KDS Redesign MADR](https://github.com/kumahq/kuma/blob/master/docs/madr/decisions/021-kds-redesign.md)  



## Why can't we just remove v1 since v2 merged?

At the v2 time, the `Delta` is something new and might have bugs. Also we should follow the compatibility between Global-CP and Zone-CP (Like a higher version Global-CP could work well with low version Zone-CP which follow the Kuma Upgrade rules `Since Kuma 1.4.x upgrades can be performed up to two minor versions`)

## Test
> multi-Zone && single-Zone, k8s && universal
- [x] Normal Test to verify if the new building cluster works well.

```text
1. Create a Multizone ENV with k8s-global, k8s-zone, universal-zone with this branch version
2. Create a `MeshTrafficPermission` policy on global and check if the resource synced to zones
3. Create a `MeshFaultInjection` policy on k8s-zone and check if the resource synced to global
4. Create a `MeshCircuitBreaker ` policy on universal-zone and check if the resource synced to global
5. Make sure no error logs in each `control-plane`
```

- [x] Upgrade Test to verify if the 2 versions previous upgrading cluster works well.

```text
1. Create a Multizone ENV with k8s-global, k8s-zone, universal-zone with Kuma `2.7.6` version
2. Upgrade global to this branch version and create a `MeshTrafficPermission` policy on global then check if the resource synced to zones
3. Upgrade k8s-zone to this branch version and create a `MeshFaultInjection` policy on k8s-zone then check if the resource synced to global
4. Upgrade universal-zone to this branch version and create a `MeshCircuitBreaker` policy on universal-zone then check if the resource synced to global
5. Stop and wait a moment to restart universal Ingress to check if the `zoningress` resources disappeared and re-created on global
6. Make sure no error logs in each `control-plane`
```

## Further thinking
Remove vendored code from `go-control-plane`: https://github.com/kumahq/kuma/issues/4926

### Checklist prior to review



- [x] [Link to relevant issue][1] as well as docs and UI issues -- https://github.com/kumahq/kuma/issues/9075
- [x] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [x] Tests (Unit test, E2E tests, manual test on universal and k8s) --
  - Don't forget `ci/` labels to run additional/fewer tests
- [x] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? -- Updated
- [x] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) -- No




[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
